src/vos/vos_query.c

```diff
@@ -255,113 +222,282 @@ _query_recx(struct open_query *query, daos_recx_t *recx, struct evt_extent *filt
+	return ent1->en_sel_ext.ex_hi > ent2->en_sel_ext.ex_hi;
+}
 
-	memset(recx, 0, sizeof(*recx));
-	/* query visible last recx */
-	opc = EVT_ITER_EMBEDDED | EVT_ITER_VISIBLE | EVT_ITER_SKIP_HOLES;
-	if (query->qt_flags & VOS_GET_MAX)
-		opc |= EVT_ITER_REVERSE;
+static bool
+lo_lt(const struct evt_entry *ent1, const struct evt_entry *ent2)
+{
+	return ent1->en_sel_ext.ex_lo < ent2->en_sel_ext.ex_lo;
+}
 
-	filter_ex.ex_lo = 0;
-	filter_ex.ex_hi = ~(uint64_t)0;
+static bool
+is_after(const struct evt_entry *ent1, const struct evt_entry *ent2)
+{
+	return ent1->en_sel_ext.ex_lo > ent2->en_sel_ext.ex_hi;
+}
 
-	return _query_recx(query, recx, &filter_ex, opc);
+static bool
+is_before(const struct evt_entry *ent1, const struct evt_entry *ent2)
+{
+	return ent1->en_sel_ext.ex_hi < ent2->en_sel_ext.ex_lo;
 }
 
-static int
-query_ec_normal_recx(struct open_query *query, daos_recx_t *recx,
-		     struct evt_extent *filter_ex, bool skip_hole)
+static void
+overlap_lo(struct evt_entry *nentry, struct evt_entry *pentry, bool *nrefresh, bool *prefresh)
 {
-	struct evt_extent tmp_ex;
-	int		  opc;
+	if (nentry->en_sel_ext.ex_lo == pentry->en_sel_ext.ex_lo) {
+		*nrefresh = true;
+		*prefresh = true;
+		return;
+	}
 
-	memset(recx, 0, sizeof(*recx));
-	/* query visible last recx */
-	opc = EVT_ITER_EMBEDDED | EVT_ITER_VISIBLE;
-	if (query->qt_flags & VOS_GET_MAX)
-		opc |= EVT_ITER_REVERSE;
+	if (nentry->en_sel_ext.ex_lo < pentry->en_sel_ext.ex_lo) {
+		*prefresh = true;
+		nentry->en_sel_ext.ex_hi = pentry->en_sel_ext.ex_lo - 1;
+		return;
+	}
 
-	if (skip_hole)
-		opc |= EVT_ITER_SKIP_HOLES;
-	else /* otherwise skip DATA */
-		opc |= EVT_ITER_SKIP_DATA;
+	pentry->en_sel_ext.ex_hi = nentry->en_sel_ext.ex_lo - 1;
+	*nrefresh = true;
+}
 
-	if (filter_ex) {
-		tmp_ex = *filter_ex;
-	} else {
-		tmp_ex.ex_lo = 0;
-		tmp_ex.ex_hi = DAOS_EC_PARITY_BIT - 1;
+static void
+overlap_hi(struct evt_entry *nentry, struct evt_entry *pentry, bool *nrefresh, bool *prefresh)
+{
+	if (nentry->en_sel_ext.ex_hi == pentry->en_sel_ext.ex_hi) {
+		*nrefresh = true;
+		*prefresh = true;
+		return;
+	}
+
+	if (nentry->en_sel_ext.ex_hi > pentry->en_sel_ext.ex_hi) {
+		*prefresh = true;
+		nentry->en_sel_ext.ex_lo = pentry->en_sel_ext.ex_hi + 1;
+		return;
+	}
+
+	pentry->en_sel_ext.ex_lo = nentry->en_sel_ext.ex_hi + 1;
+	*nrefresh = true;
+}
+
+static inline void
+ent2recx(daos_recx_t *recx, const struct evt_entry *ent)
+{
+	recx->rx_idx = ent->en_sel_ext.ex_lo;
+	recx->rx_nr = ent->en_sel_ext.ex_hi - ent->en_sel_ext.ex_lo + 1;
+	D_DEBUG(DB_TRACE, "ec_recx size is "DF_X64"\n", recx->rx_idx + recx->rx_nr);
+}
+
+static bool
+find_answer(struct evt_entry *pentry, struct evt_entry *nentry, daos_recx_t *recx, bool *nrefresh,
+	    bool *prefresh, check_func start_chk, check_func after_chk,
+	    void (*handle_overlap)(struct evt_entry *, struct evt_entry *, bool *, bool *))
+{
+	if (start_chk(pentry, nentry)) {
+		/** Use the adjusted parity extent */
+		ent2recx(recx, pentry);
```

Comment:
Should the returnd recx being adjusted if the 'pentry' and 'nentry' are overlapping?

I'm not sure what the query recx API will be used for, if it's only used for array size query, I think we are good. if the query recx is used for querying the specific logical extent, looks the returned extent needs be adjusted?

Reply:
It's used for max or min size query.  We only care about the logical size here.